helsingborg-stad/app-mitt-helsingborg
# Feat/encryption with coapplicant symmetric key

NOTE: WiP - putting as draft as a form of documentation of the progress.

## Explain the changes you’ve made

* Rewritten encryption system (`EncryptionService` and `EncryptionHelper`).
* Added basic `StorageService` tests.
* Added basic encryption tests.
* Added an extra helper function to perform deep merging.

## Explain why these changes are made

Encryption codebase is rewritten to clean it up and to alleviate further development:
* All functions are now fully TypeScript typed. 
* `EncryptionService` is now a generic file without any 1st-party dependencies except for `StorageService`; it could essentially be lifted out and inserted into a new project without major rework.
* `EncryptionHelper` is now the wrapper around `EncryptionService` and is generally the file that should be used directly. It contains the functionality for encrypting/decrypting forms, exchanging symmetric key values etc.
* All cryptographically-relevant numbers are now coerced to `CryptoNumber` and `SerializedCryptoNumber` (custom types). 
    * The intent is to be able to change the underlying type to `BigInt` in the future, without much refactoring. `SerializedCryptoNumber` is a JSON-compliant representation of the number (intended to be a hex-string for `BigInt`-serialization).
* Under the hood encryption now uses generic strings (referred to as "ID" or "Unique ID") as storage keys.
  * This is in order to decouple encryption from the previous `User.personalNumber` paradigm. Further up the chain the unique IDs still usually resolve to the user's personal number.
* All generic functions are now functional and don't modify any input data.
  * The consequence is that some extra merging is required for a few operations; `EncryptionHelper` usually handles this.

## Explain your solution

* `Case` and `Encryption` type files exist under `source/types`.
* Regular private AES encryption (single-applicant) works as before, but with slightly changed function names for clarification.
* The encryption stories tests are updated to reflect codebase changes.

## How to test the changes?

Note: when reviewing the code it is probably best to ignore "changes" for the encryption files and instead consider them as entirely new files as they've been largely rewritten from scratch.

1. Checkout this branch.
    * Note: underlying storage keys are changed so any previous cases will be un-decryptable. It is recommended to start from a clean state.
2. Verify that private AES encryption works as before (e.g. by starting a case, fill in a few answers, close the form, close the app, re-open the app and verify that the form can still be opened and the answers are there).

## Was this feature tested in the following environments?
- [] Storybook on a iOS device/simulator.
- [] Building the Application on a Android device/simulator.

## Screenshots (optional)


## Anything else? (optional)

Some files (type definitions and encryption files) used #503 locally for linting and formatting, so any further edits to these files without using #503 might format the files in a different way and show false-positive ESLint errors.
